pkg/cmd/microbench-ci/run.go

```
@@ -77,29 +77,48 @@ func (b *Benchmark) run() error {
 			return err
 		}
 	}
-
-	log.Printf("Running benchmark %q for %d iterations", b.Name, b.Count)
-	for i := 0; i < b.Count; i++ {
-		for _, revision := range []Revision{New, Old} {
-			log.Printf("%s binary iteration (%d out of %d)",
-				revision, i+1, b.Count,
-			)
-			err := b.runIteration(revision, fmt.Sprintf("%d", i+1))
-			if err != nil {
-				return err
+	status := NoChange
+	tries := 0
+	for retries := 0; retries < b.Retries; retries++ {
+		log.Printf("Running benchmark %q for %d iterations (attempt %d)", b.Name, b.Count, retries+1)
+		for i := 1; i <= b.Count; i++ {
+			for _, revision := range []Revision{New, Old} {
+				log.Printf("%s binary iteration (%d out of %d)",
+					revision, i, b.Count,
+				)
+				err := b.runIteration(revision, fmt.Sprintf("%d", (tries*b.Count)+i))
+				if err != nil {
+					return err
+				}
 			}
 		}
-	}
+		tries++
 
-	// Write failure marker file if the benchmark regressed.
-	compareResult, err := b.compare()
-	if err != nil {
-		return err
+		compareResult, err := b.compare(b.Count)
+		if err != nil {
+			return err
+		}
+		compareStatus := compareResult.top()
+		// If the benchmark shows no change, we can stop running additional iterations
+		if compareStatus == NoChange {
+			break
+		}
+		// Track the most significant change we've seen (Regressed > Improved > NoChange)
```

Shouldn't this be the other way around? If only one of the three runs showed a regression while the other two were `NoChange`, we want it to be `NoChange`?

---

Although that leaves an edge case we may or may not care about. If we see two regressions and one improved, it would be marked as improved even though it should be `NoChange`. Perhaps that's rare and seems not that consequential.

---

Shouldn't we be looking at the amplitude of a regression? If the regression is just due to noise, then retries don't really help.

---

> Shouldn't we be looking at the amplitude of a regression? If the regression is just due to noise, then retries don't really help.

We rely one the p-value to decide if it's a regression (big or small), the confidence is what tells us if we should try and reproduce again. If we're able to reproduce this 3 times the chances of it being random chance is extremely low.

> Shouldn't this be the other way around?

There is a case here I didn't think about which I need to address and that is if it flipped from Regression to Improved, which would also throw out the results.

The idea behind this is to only keep running if we keep seeing a regression, the moment we don't the results are deemed insignificant.

In other words the only scenario we deem regression worthy is 3 consecutive runs with regressions. Any other scenario fails early so that is doesn't waste any more CI time.

---

> if only one of the three runs showed a regression while the other two were NoChange, we want it to be NoChange? i.e. if we marked all regressions as such with only one occurrence of a regression, why retry?

Good catch, the comment there is silly (I'll update it), but yes there's a small bug here that should be addressed, if it changes from regressed <-> improved, then the results will be wrong; i.e., only 2 scenarios lead to a report (positive): [regressed x 3] or [improved x 3] - not a mix. And any `NoChange` should short circuit and mark the whole set as insignificant.

---

> In other words the only scenario we deem regression worthy is 3 consecutive runs with regressions. Any other scenario fails early so that it doesn't waste any more CI time.

That makes sense! Assuming the runs are totally independent, we can derive the p-value for each independent run and compute the resulting one, e.g., using Fisher's method [1]. But we don't have to get too fancy; intuitively, it makes sense to call it a potential regression.

[1] https://en.wikipedia.org/wiki/Fisher%27s_method